### PR TITLE
Mapgen Watershed development

### PR DESCRIPTION
**Pull request by paramat**
WIP. Terrain is still being tuned, large scale base terrain shape is being reconsidered, a few other things to do, but in a reasonable state and usable.

As this mapgen uses a new node and needs additional mapgen aliases you need to use this mod with it https://github.com/paramat/volcanic
Uses the Biome API just like all other non-mgv6 mapgens so will be compatible with biome system mods like Ethereal.

(Tenplus1: it doesn't have that feature, sorry, but maybe i'll add floatlands).

### CODE REVIEW

File: src/mapgen/mapgen_watershed.h [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License along
+with this program; if not, write to the Free Software Foundation, Inc.,
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#ifndef MAPGEN_WATERSHED_HEADER
```
Comment (nerzhul) [inline src/mapgen/mapgen_watershed.h]: please use pragma once

Comment (paramat) [reply src/mapgen/mapgen_watershed.h]: Will do.

**Comment by paramat** [comment]
Will also add to LINT whitelist.

**Comment by paramat** [comment]
Build will fail due to LINT but will whitelist it closer to merge, some of the LINT suggestions may be useful.

File: src/mapgen/mapgen_watershed.h (hunk 1 above)
Comment (paramat) [reply src/mapgen/mapgen_watershed.h]: Done.

**Comment by kilbith** [comment]
No :-1: One simple mapgen is enough, no need to add new fancy ones.

Joke aside, we should deal with your PR like you with the new proposed features, particularly on the graphics. This isn't fair-play otherwise.

**Comment by jastevenson303** [comment]
To be fair, and to save paramat the typing, he is currently active to maintain it.

**Comment by paramat** [comment]
kilbith likes to take any opportunity to pester me.

> we should deal with your PR

'we' is not the right word, you're not a MT dev.

**Comment by kilbith** [comment]
> you're not a MT dev.

Yep, fortunatety. People shouldn't consider you as a core-dev either.

**Comment by SmallJoker** [comment]
How the mapgen currently looks (seed = "watershed") with coordinates around 0,0,0 and 500,0,500:
screenshot_20171215_175011
screenshot_20171215_175541

**Comment by pilino1234** [comment]
Terrain looks like a first draft of a mapgen? 👎

**Comment by paramat** [comment]
pilino1234 is another who likes to take opportunities to pester me, and tends to turn up alongside kilbith, or when i'm in argument. So i don't consider their comments reasonable.

**Comment by pilino1234** [comment]
> So i don't consider their comments reasonable.

your loss. But my point still stands 😄

**Comment by paramat** [comment]
No, because i know that there are many gentle slopes in it as well as lot of variation and rough areas.
So did you try the mapgen and travel many 1000s of nodes or judge on 2 screenshots?

**Comment by benrob0329** [comment]
I think the mapgen looks interesting, but also a bit odd. I feel like the huge vertical cliffs here, combined with the huge caves of mgv6, the mountains of mgv7, and the rivers of valleys would make one awesome mapgen, rather than 4 or 5 good ones.

**Comment by kilbith** [comment]
> pilino1234 is another who likes to take opportunities to pester me

You never wondered why you arise such hostility against you in the community? Also pilino1234 is not my henchman, he comes up with some constructive critics and you take it in a personal way. Poor guy, come to London in January if you have guts.

**Comment by paramat** [comment]
>  Poor guy, come to London in January if you have guts.

Another physical threat.

**Comment by kilbith** [comment]
No, an invitation to discuss in face to face. I don't know how do you interpret that as a "physical threat", I never did that.

**Comment by benrob0329** [comment]
\*sigh\*

**Comment by paramat** [comment]
It looks like a physical threat to most, that's all that matters.

**Comment by rubenwardy** [comment]
Both @kilbith's and @pilino1234's first comments are valid comments, and valid criticisms. What exactly does this mapgen offer that others don't? Why did you open a PR when it's half complete - what do you want feedback on?

Whilst mapgens are mostly self-contained - don't touch other code parts much - it's still something extra to maintain, so should offer something new. For example, better erosion mechanics.

**Comment by masrourmouad** [comment]
Nice job Paramat , but i think that the mapgen should be  self-contained

**Comment by paramat** [comment]
rubenwardy, they are not valid comments.

> we should deal with your PR like you with the new proposed features, particularly on the graphics.

This is obviously taking an opportunity to pester, as it implies my reviews of features are somehow unreasonable, as kilbith believes, so suggests that reviews of this should be unreasonable, which is what pilino1234 then did, and to an extent you did too with your comment.

> terrain looks like a first draft of a mapgen?

We have cliff terrain just like this in mgv7. The first screenshot shows cliff terrain running int a mountain, neither screenshot show anything less developed than what you see in other mapgens.

> slopes look way to steep to be climbable

These are cliffs, mgv6 and mgv7 also have unclimbable cliff slopes.
The mountain is steep because it is a mountain, mgv7, valleys and carpathian all have steeper mountains than this.
The mapgen has a variety of terrain, there's more than cliffs and steep mountains.

> and it looks very unnatural

It's actually designed to be more natural than many mapgens, however i could add another octave of noise to the cliff terrain shape, it's possibly a little too rounded, if that's what you meant.
It seems pilino1234 only looked at 2 screenshots, didn't try the mapgen, then made an effort to be as negative and unreasonable as possible.

Rubenwardy your comment gives the impression you are joining in and being overcritical.

This offers things other mapgens don't, more realistic rivers, canyons, volcanos and a mapping feature that allows you to shrink the world to any scale to use as a 3D map.
However also it combines the best features of other mapgens, which is exactly what c55 and hmmmm had in mind for new mapgens.
It's not half finished, it just needs tuning and possibly floatlands adding. The reason i opened it now is because i have been working on it for a long time and realised i could continue tweaking it forever, opening a PR helps me to focus on it by giving me a deadline of 0.5.0.
I don't need any feedback, but reasonable feedback is welcome and will be considered.

When the creators of mgvalleys and mgcarpathian offered their mapgens, i and others were very 'hands-off', allowing them to create what they want and accepting it as long as it was liked and met minimum quality standards. No-one became overly concerned with if it added something new, and gave the authors time. Now that i make one, people are being overcritical and overquestioning, please treat me how i treated the creators of mgvalleys and mgcarpathian. As the magen dev, i will treat myself the way i treated them.

> but i think that the mapgen should be self-contained

Why? It's designed to use the Biome API and functions of class MapgenBasic, cavegen, dungeongen etc. so can't be. Our mapgen architecture is all about sharing common functions. If i twas slef-contained no-one could use it with their biome mods or games.

**Comment by rubenwardy** [comment]
You've just refuted their argument and explained what this PR offers, that's what I was looking for

**Comment by rubenwardy** [comment]
I will say that the behaviour of kilbith in this thread was completely unprofessional and uncalled for

**Comment by paramat** [comment]
Although, don't get your hopes up for the volcanos, you may understandably be expecting black conical mountains with a central vent, but they are difficult to implement so are essentially 'magma vents', deep pits in the ground surrounded by volcanic rock, but i'll try to improve them.

There is also a lot of flat terrain as has been much requested, for building on.

**Comment by paramat** [comment]
screenshot_20171217_090459

screenshot_20171217_090848

screenshot_20171217_091409

^ Obsidian dungeon in opposite wall

screenshot_20171217_091516

screenshot_20171217_091946

**Comment by paramat** [comment]
screenshot_20171217_103924

^ 3D map feature with map_scale = 16. Found a nice evil looking mountain, multiplied co-ords by 16:

screenshot_20171217_105130

^ 1049 nodes high

**Comment by paramat** [comment]
screenshot_20171217_110618

^ Canyon meeting the sea

**Comment by kilbith** [comment]
> I will say that the behaviour of kilbith in this thread was completely unprofessional

Minetest is not a professional organization.

**Comment by HummusSamurai** [comment]
I rarely get to comment here because of inactivity, but I must say this is very pretty!

I _adore_ the flat mountain tops and the evil looking mountains and the water bodies.

I hope it gets merged soon.

**Comment by pilino1234** [comment]
> then made an effort to be as negative and unreasonable as possible.

it's your job as the author of this pr to convince people that adding yet another mapgen is necessary (by providing screenshots etc.)
what does this new mapgen provide that other mapgens don't provide by tweaking their "essential, non-negotiable" mapgen parameters that there are so many (and so little user-friendly docs) of? Currently, all the mapgens feel like similar iterations of the same thing.

> only looked at 2 screenshots, didn't try the mapgen
Yup, when i wrote the comment i had only seen the two screenshots that Krock had provided, because there was nothing else available.
I've tested the mapgen for quite a bit now, and find that especially the jungle terrains extend for extremely long distances. considering that jungle is a pain to travel through as a player, i suggest reducing the size of the jungles. The actual terrain looks nice enough, but I feel it's too similar to previous mapgens to warrant adding yet another one.

**Comment by paramat** [comment]
Obsidian dungeons built through volcanos.

screenshot_20171217_143719

screenshot_20171217_143942

screenshot_20171217_145538

**Comment by Ezhh** [comment]
Looking forward to wandering round those volcanoes. :)

**Comment by srifqi** [comment]
So, we have obsidian volcanoes? Woah!

**Comment by catagent101** [comment]
I really like the plateaus, they look like great places to build a home on, since they have a nice view of the area below. The volcanoes are also impressive.

**Comment by C1ffisme** [comment]
I actually think this mapgen looks pretty cool, although I agree with the people that say that this is a little too steep. (Although a part of me wants it to be as steep as it currently is, because that's part of the "wow" factor of this mapgen.)

**Comment by paramat** [comment]
The volcanos are lined with a new node 'volcanic rock'.
The screenshots above only show steep terrain, there are more gentle slopes elsewhere (that admittedly need tuning).

**Comment by paramat** [comment]
Removed screenshots.
I was hoping opening this PR would help me to complete the mapgen but i realise there is still much to do. I'll reopen in a new PR later.

**Comment by paramat** [comment]
mgw2s30tanhf

@Ezhh very early continent experiment. This map is 2000x2000 nodes but the generation is scaled down by a factor of 30 (a useful feature of this mapgen used for creating 3D maps), so you are seeing an entire MT world here.
This is not ideal of course but it is difficult to tune, will keep experimenting.

**Comment by paramat** [comment]
wbt2

A little clearer, this is the base terrain without valleys, plateaus or mountains. A whole MT world again.

**Comment by paramat** [comment]
mgw2s30d

Improving ..

**Comment by paramat** [comment]
mgw2s30_ss_2p0

@Ezhh again scaled down to show an entire MT world. This time sandstone shows mountainous areas above y = 128 (cloud level).

**Comment by paramat** [comment]
w2abs7

A whole world again, but with a different use of continent noise that creates a little more isolation of continents, as opposed to the more connected weblike distribution shown before.
@Ezhh

**Pull request closed without merging**